Clarify that console.assert should log even without formatting string

https://console.spec.whatwg.org/commit-snapshots/36c3ee56b8471c82ec5ea88dcbaf223235b22a3e/#assert

> If condition is false, perform Logger\("error", data\)\.

A call like `console.assert(false)` would then call `Logger("error", [])`, but the implementation of `Logger` states as the first item:

> 1. If `args` is empty, abort these steps.

I would expect the behavior to log something though - and the participants in the discussion around #56 seem to make the same assumption.

When calling `console.assert(false)` in various browsers, something is logged in that case:

- Chrome prints `Assertion failed: console.assert` with level *error*
- Safari prints `Assertion Failed` with level *error*
- Firefox prints an empty line with level *error*
- Edge prints `undefined` with level *error*
- Node throws an `AssertionError: false == true` (`false` is static, same result for `console.assert(0)`)

I'd propose to add a section that clarifies this:

1. If condition is *true*, abort these steps.
2. If *data* is not empty, perform Logger("error", *data*). Abort these steps.
3. Let *message* be the string "Assertion failed".
4. Perform Logger("error", [*message*])

Should we make a clear recommendation of what the String in (3) should look like? Or leave it open to implementers, e.g. just say "a string, indicating that some assertion failed (e.g. ...)"?

---

For reference, calling `console.assert(false, "foo")` in various browsers:

- Chrome prints `Assertion failed: foo` with level *error*
- Safari prints `Assertion Failed: foo` with level *error*
- Firefox prints `foo` with level *error*
- Edge prints `foo` with level *error*
- Node throws an `AssertionError: foo`

EDIT: Didn't have access to Edge, added @domenic's findings.

--- a/index.bs
+++ b/index.bs
@@ -231,7 +231,7 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
   1. If _data_ is empty, append _message_ to _data_.
   1. Otherwise, implementations should perform these substeps:
     1. Let _first_ be the first element of _data_.
-    1. If _first_ is not a string, prepend _message_ to _data_. Abort these substeps.
+    1. If Type(_first_) is not String, prepend _message_ to _data_. Abort these substeps.
     1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 (SPACE), and _first_.
     1. Set _data[0]_ to _concat_.
   1. Perform Logger("error", _data_).

--- a/index.bs
+++ b/index.bs
@@ -232,7 +232,7 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
   1. Otherwise, implementations should perform these substeps:
     1. Let _first_ be the first element of _data_.
     1. If Type(_first_) is not String, prepend _message_ to _data_. Abort these substeps.
-    1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 (SPACE), and _first_.
+    1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 SPACE, and _first_.
     1. Set _data[0]_ to _concat_.
   1. Perform Logger("error", _data_).
 </emu-alg>

--- a/index.bs
+++ b/index.bs
@@ -223,7 +223,7 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
 
 <h3 id="logging">Logging methods</h3>
 
-<h4 id="assert" oldids="assert-condition-data,dom-console-assert" method for="console">assert(<var>condition</var>, ...<var>data</var>)</h4>
+<h4 id="assert" oldids="assert-condition-data,dom-console-assert" method for="console">assert(<var ignore=''>condition</var>, ...<var>data</var>)</h4>
 
 <emu-alg>
   1. If _condition_ is true, abort these steps.

--- a/index.bs
+++ b/index.bs
@@ -223,7 +223,7 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
 
 <h3 id="logging">Logging methods</h3>
 
-<h4 id="assert" oldids="assert-condition-data,dom-console-assert" method for="console">assert(<var ignore=''>condition</var>, ...<var>data</var>)</h4>
+<h4 id="assert" oldids="assert-condition-data,dom-console-assert" method for="console">assert(<var>condition</var>, ...<var>data</var>)</h4>
 
 <emu-alg>
   1. If _condition_ is true, abort these steps.
@@ -231,7 +231,7 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
   1. If _data_ is empty, append _message_ to _data_.
   1. Otherwise, implementations should perform these substeps:
     1. Let _first_ be the first element of _data_.
-    1. If _first_ is not a string, prepend _data_ with _message_. Abort these substeps.
+    1. If _first_ is not a string, prepend _message_ to data_. Abort these substeps.
     1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 (SPACE), and _first_.
     1. Replace the first element of _data_ with _concat_.
   1. Perform Logger("error", _data_).

--- a/index.bs
+++ b/index.bs
@@ -233,7 +233,7 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
     1. Let _first_ be the first element of _data_.
     1. If _first_ is not a string, prepend _message_ to _data_. Abort these substeps.
     1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 (SPACE), and _first_.
-    1. Replace the first element of _data_ with _concat_.
+    1. Set _data[0]_ to _concat_.
   1. Perform Logger("error", _data_).
 </emu-alg>
 

--- a/index.bs
+++ b/index.bs
@@ -231,10 +231,9 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
   1. If _data_ is empty, append _message_ to _data_.
   1. Otherwise, implementations should perform these substeps:
     1. Let _first_ be the first element of _data_.
-    1. If _first_ is not a string, prepend _data_ with _message_.
-    1. Otherwise, perform these substeps:
-      1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 (SPACE), and _first_.
-      2. Replace the first element of _data_ with _concat_.
+    1. If _first_ is not a string, prepend _data_ with _message_. Abort these substeps.
+    1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 (SPACE), and _first_.
+    1. Replace the first element of _data_ with _concat_.
   1. Perform Logger("error", _data_).
 </emu-alg>
 

--- a/index.bs
+++ b/index.bs
@@ -231,7 +231,7 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
   1. If _data_ is empty, append _message_ to _data_.
   1. Otherwise, implementations should perform these substeps:
     1. Let _first_ be the first element of _data_.
-    1. If _first_ is not a string, prepend _message_ to data_. Abort these substeps.
+    1. If _first_ is not a string, prepend _message_ to _data_. Abort these substeps.
     1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 (SPACE), and _first_.
     1. Replace the first element of _data_ with _concat_.
   1. Perform Logger("error", _data_).

--- a/index.bs
+++ b/index.bs
@@ -223,9 +223,20 @@ its \[[Prototype]] an empty object, created as if by ObjectCreate(%ObjectPrototy
 
 <h3 id="logging">Logging methods</h3>
 
-<h4 id="assert" oldids="assert-condition-data,dom-console-assert" method for="console">assert(<var>condition</var>, ...<var>data</var>)</h4>
+<h4 id="assert" oldids="assert-condition-data,dom-console-assert" method for="console">assert(<var ignore=''>condition</var>, ...<var>data</var>)</h4>
 
-If <var>condition</var> is false, perform Logger("error", <var>data</var>).
+<emu-alg>
+  1. If _condition_ is true, abort these steps.
+  1. Let _message_ be a string without any formatting specifiers indicating generically an assertion failure (such as "Assertion failed").
+  1. If _data_ is empty, append _message_ to _data_.
+  1. Otherwise, implementations should perform these substeps:
+    1. Let _first_ be the first element of _data_.
+    1. If _first_ is not a string, prepend _data_ with _message_.
+    1. Otherwise, perform these substeps:
+      1. Let _concat_ be the concatenation of _message_, U+003A COLON (:), U+0020 (SPACE), and _first_.
+      2. Replace the first element of _data_ with _concat_.
+  1. Perform Logger("error", _data_).
+</emu-alg>
 
 <h4 id="clear" oldids="dom-console-clear" method for="console">clear()</h4>
 

